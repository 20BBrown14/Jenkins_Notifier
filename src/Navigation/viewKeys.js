export const VK_ADD_NEW_JOB = 'add_new_job';
export const VK_JOB_STATUS = 'job_status';
