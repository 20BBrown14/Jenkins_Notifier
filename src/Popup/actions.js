
/** @const action for validating repo url in state */
export const A_VALIDATE_REPO_URL = 'validate_repo_url';

/** @const action for when job input field data changed */
export const A_JOB_INPUT_DATA_CHANGED = 'job_input_data_changed';

/**
 * Action creator for A_JOB_INPUT_DATA_CHANGED
 */
export const jobInputDataChanged = newValue => (
  {
    type: A_JOB_INPUT_DATA_CHANGED,
    data: { newValue },
  }
);
