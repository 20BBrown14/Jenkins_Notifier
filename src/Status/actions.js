/** @const action for when Add New Repo button is clicked */
export const A_ADD_NEW_REPO_CLICKED = 'add_new_repo_clicked';
/** @const action to remove repo from list */
export const A_REMOVE_REPO = 'remove_repo';
/** @const action to view a repo's jobs */
export const A_VIEW_JOBS_CLICKED = 'view_jobs_clicked';
/** @const action to remove job from repo */
export const A_REMOVE_JOB_CLICKED = 'remove_job_clicked';
/** @const action to go back to repo view from jobs view */
export const A_GO_BACK_TO_REPO_VIEW = 'go_back_to_repo_view';


export const removeRepoAction = repoToRemove => (
  {
    type: A_REMOVE_REPO,
    data: { repoToRemove },
  }
);


export const viewJobsClickedAction = repoToView => (
  {
    type: A_VIEW_JOBS_CLICKED,
    data: { repoToView },
  }
);

export const removeJobAction = (jobToRemove, repo) => (
  {
    type: A_REMOVE_JOB_CLICKED,
    data: { jobToRemove, repo },
  }
);
