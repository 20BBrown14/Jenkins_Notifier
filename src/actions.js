/** @const action to refresh repo */
export const A_REFRESH_REPO = 'refresh_repo';
/** @const action to indicate a repo has been refreshed */
export const A_REPO_IS_REFRESHED = 'repo_is_refreshed';

export const refreshRepoAction = url => (
  {
    type: A_REFRESH_REPO,
    data: { url },
  }
);

export const repoIsRefreshedAction = (data, repoName) => (
  {
    type: A_REPO_IS_REFRESHED,
    data: { data, repoName },
  }
);
