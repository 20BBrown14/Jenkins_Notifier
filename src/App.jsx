import React, { Component } from 'react';
import createSagaMiddleware from 'redux-saga';
import { combineReducers, createStore, compose, applyMiddleware } from 'redux';
import Application from 'terra-clinical-application';
import { Provider } from 'react-redux';
import './App.css';
import Popup from './Popup/PopupContainer';


const sagaMiddleware = createSagaMiddleware();

const rootReducer = combineReducers(Object.assign({}, null));

/** @const redux store */
const store = createStore(rootReducer, compose(applyMiddleware(sagaMiddleware)));

[].map(saga => (sagaMiddleware.run(saga)));

/* eslint-disable-next-line react/prefer-stateless-function */
class App extends Component {
  render() {
    return (
      <Provider store={store}>
        <Application>
          <Popup />
        </Application>
      </Provider>
    );
  }
}

export default App;
