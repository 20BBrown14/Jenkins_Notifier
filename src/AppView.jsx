import React from 'react';
import PropTypes from 'prop-types';
import AddRepo from './AddRepo/AddRepoContainer';
import { VK_ADD_NEW_REPO, VK_REPOS } from './Navigation/viewKeys';
import Status from './Status/StatusContainer';

const propTypes = {
  /** View key for the current view */
  viewKey: PropTypes.string.isRequired,
  /** Repo information stored */
  repos: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
};

export const AppView = (props) => {
  const {
    viewKey,
    repos,
  } = props;

  let content;
  switch (viewKey) {
    case VK_REPOS:
      content = (
        <Status
          repos={repos}
        />
      );
      break;
    case VK_ADD_NEW_REPO:
      content = (<AddRepo />);
      break;
    default:
      content = undefined;
      break;
  }

  return (
    <div>
      {content}
    </div>
  );
};

AppView.propTypes = propTypes;
export default AppView;
