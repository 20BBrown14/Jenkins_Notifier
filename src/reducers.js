import { A_VALID_REPO_URL, A_INFORMATION_CONFIRMED, A_CANCEL_CLICKED } from './AddRepo/actions';
import { VK_ADD_NEW_REPO, VK_REPOS } from './Navigation/viewKeys';
import { A_ADD_NEW_REPO_CLICKED, A_REMOVE_REPO, A_REMOVE_JOB_CLICKED, A_GO_BACK_TO_REPO_VIEW } from './Status/actions';
import { A_LOAD_APP_STATE } from './actions';

export const APP_STATE = 'app_state';

const INITIAL_STATE = {

  /* repos: {
    patientKioskJS: {
      URL: 'https://jenkins.cerner.com/ion/job/Revenue%20Cycle%20Patient%20Kiosk/job/patient-kiosk-js/',
      jobs: {
        0: {
          color: 'blue',
          name: 'dev',
          url: 'https://jenkins.cerner.com/ion/job/Revenue%20Cycle%20Patient%20Kiosk/job/patient-kiosk-js/job/dev/',
        },
        1: {
          color: 'red',
          name: 'cheese',
          url: 'https://jenkins.cerner.com/ion/job/Revenue%20Cycle%20Patient%20Kiosk/job/patient-kiosk-js/job/feature%252FREVCYCSCH-3884/'
        },
      },
    },
    kioskServer: {
      URL: 'https://jenkins.cerner.com/ion/job/Revenue%20Cycle%20Patient%20Kiosk/job/kiosk_server/',
      jobs: {
        0: {
          color: 'blue',
          name: 'dev',
          url: 'https://jenkins.cerner.com/ion/job/Revenue%20Cycle%20Patient%20Kiosk/job/patient-kiosk-js/job/dev/',
        },
        1: {
          color: 'red',
          name: 'cheese',
          url: 'https://jenkins.cerner.com/ion/job/Revenue%20Cycle%20Patient%20Kiosk/job/patient-kiosk-js/job/feature%252FREVCYCSCH-3884/'
        },
      },
    },
  }, */

  repos: undefined,
  viewKey: VK_REPOS,
};

const reduceValidRepoURL = (state, action) => (
  {
    ...state,
    jsonData: action.data.jsonData,
  }
);

const reduceInformationConfirmed = (state, action) => {
  const repoInformation = state.jsonData;
  const repoName = action.data.repoName || repoInformation.displayName;
  const newState = { ...state };
  /* eslint-disable no-restricted-syntax */
  if (newState.repos) {
    for (const name in newState.repos) {
      if (newState.repos[name].URL === repoInformation.url) {
        delete newState.repos[name];
      }
    }
  }
  /* eslint-enable no-restricted-syntax */
  return {
    ...newState,
    repos: {
      ...newState.repos,
      [repoName]: {
        URL: repoInformation.url,
        jobs: repoInformation.jobs,
      },
    },
  };
};

const reduceRemoveRepo = (state, action) => {
  const { repoToRemove } = action.data;
  const newState = { ...state };
  delete newState.repos[repoToRemove];
  return { ...newState, repos: { ...newState.repos } };
};

const reduceRemoveJob = (state, action) => {
  const { jobToRemove, repo } = action.data;
  const newState = { ...state };
  delete newState.repos[repo].jobs[jobToRemove];
  if (Object.keys(newState.repos[repo].jobs).length === 0) {
    const reposToReturn = { ...newState.repos };
    reposToReturn[repo].jobs = undefined;
    return { ...newState, repos: reposToReturn };
  }
  return { ...newState, repos: { ...newState.repos } };
};


/**
 * Define App reducers
 * @param {object} state - State with default value of INITIAL_STATE
 * @param {string} state.repoURL - Current value of the repoURL field.
 * @param {object} action - The action being performed
 * @param {string} action.type - The type of action being performed
 * @param {object} action.data - The data related to the action
 * @returns {object} updated state. If no action is handled returns original state.
 */
const reduceApp = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case A_INFORMATION_CONFIRMED:
      return reduceInformationConfirmed(state, action);
    case A_VALID_REPO_URL:
      return reduceValidRepoURL(state, action);
    case A_ADD_NEW_REPO_CLICKED:
      return { ...state, viewKey: VK_ADD_NEW_REPO };
    case A_CANCEL_CLICKED:
      return { ...state, viewKey: VK_REPOS };
    case A_REMOVE_REPO:
      return reduceRemoveRepo(state, action);
    case A_REMOVE_JOB_CLICKED:
      return reduceRemoveJob(state, action);
    case A_GO_BACK_TO_REPO_VIEW:
      return { ...state, viewKey: VK_REPOS };
    case A_LOAD_APP_STATE:
      return { ...state, repos: action.data.repos };
    default:
      return state;
  }
};

export default { [APP_STATE]: reduceApp };
