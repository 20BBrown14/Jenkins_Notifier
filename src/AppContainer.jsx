import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { APP_STATE } from './reducers';
import View from './AppView';
import { refreshRepoAction } from './actions';
import { googleStorageReposGet } from './modules/googleHelpers';

const propTypes = {
  /** View key for the current view */
  viewKey: PropTypes.string.isRequired,
  /** Stored repo information in the state */
  repos: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  /** Function to dispatch refresh repo action */
  refreshRepo: PropTypes.func.isRequired,
};

export class AppContainer extends React.Component {
  componentWillMount() {
    const {
      refreshRepo,
    } = this.props;
    googleStorageReposGet(refreshRepo);
  }

  render() {
    const {
      viewKey,
      repos,
    } = this.props;

    return (
      <div>
        <View
          viewKey={viewKey}
          repos={repos}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  viewKey: state[APP_STATE].viewKey,
  repos: state[APP_STATE].repos,
});

export const mapDispatchToProps = dispatch => ({
  refreshRepo: (url, repoName) => {
    dispatch(refreshRepoAction(url, repoName));
  },
});

AppContainer.propTypes = propTypes;
export default connect(mapStateToProps, mapDispatchToProps)(AppContainer);
