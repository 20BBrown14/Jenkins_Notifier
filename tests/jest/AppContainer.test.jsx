import React from 'react';
import { AppContainer, mapDispatchToProps } from '../../src/AppContainer';
import { VK_ADD_NEW_REPO, VK_REPOS } from '../../src/Navigation/viewKeys';
import { A_LOAD_APP_STATE } from '../../src/actions';

describe('AppContainer', () => {
  afterEach(() => {
    global.chrome.storage.sync.get = () => {};
  });
  describe('Initialization', () => {
    it('should render a default view', () => {
      const testContainer = (mount(
        <AppContainer
          viewKey="someViewKey"
          repos={{}}
          loadAppState={() => {}}
        />,
      ));
      expect(testContainer).toMatchSnapshot();
    });
    it('should trigger componentwillmount life cycle event', () => {
      const loadAppStateMock = jest.fn();
      global.chrome.storage.sync.get = jest.fn((param1, someFunction) => {
        someFunction({ repos: {} });
      });
      const testContainer = (mount(
        <AppContainer
          viewKey="someViewKey"
          repos={{}}
          loadAppState={loadAppStateMock}
        />,
      ));
      expect(testContainer).toMatchSnapshot();
      expect(loadAppStateMock).toHaveBeenCalled();
      expect(loadAppStateMock).toHaveBeenCalledTimes(1);
      expect(global.chrome.storage.sync.get).toHaveBeenCalled();
    });
  });
  describe('When view key is VK_REPOS', () => {
    it('should render a status view', () => {
      const wrapper = (shallow(
        <AppContainer
          viewKey={VK_REPOS}
          repos={{}}
          loadAppState={() => {}}
        />,
      ));
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('View')).toBeDefined();
      expect(wrapper.find('View').find('Status')).toBeDefined();
    });
  });
  describe('When view key is VK_ADD_NEW_REPO', () => {
    it('should render an add new repo view', () => {
      const wrapper = (shallow(
        <AppContainer
          viewKey={VK_ADD_NEW_REPO}
          repos={{}}
          loadAppState={() => {}}
        />,
      ));
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('View')).toBeDefined();
      expect(wrapper.find('View').find('AddRepo')).toBeDefined();
    });
  });
  describe('Component did update', () => {
    it('should set repos in storage', () => {
      global.chrome.storage.sync.get = jest.fn((param1, someFunction) => {
        someFunction({ repos: 'someRepo' });
      });
      global.chrome.storage.sync.set = jest.fn((param1, param2) => {
        param2();
        expect(param1).toEqual({ repos: {} });
      });
      const wrapper = shallow(
        <AppContainer
          viewKey="someViewKey"
          repos={null}
          loadAppState={() => {}}
        />,
      );
      wrapper.setProps({ viewKey: 'someOtherViewKey' });
      expect(wrapper).toMatchSnapshot();
      expect(global.chrome.storage.sync.set).toHaveBeenCalled();
      expect(global.chrome.storage.sync.set).toHaveBeenCalledTimes(1);
    });
  });
  describe('action dispatch', () => {
    let mockDispatch;
    afterEach(() => {
      expect(mockDispatch).toHaveBeenCalled();
      mockDispatch = undefined;
    });
    it('should dispatch load app state action', () => {
      mockDispatch = jest.fn((action) => {
        expect(action.type).toEqual(A_LOAD_APP_STATE);
      });
      const dispatchProps = mapDispatchToProps(mockDispatch);
      dispatchProps.loadAppState();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });
  });
});
