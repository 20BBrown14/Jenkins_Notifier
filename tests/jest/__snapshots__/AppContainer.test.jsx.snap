// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer Initialization should render a default view 1`] = `
<AppContainer
  refreshRepo={[Function]}
  repos={Object {}}
  viewKey="someViewKey"
>
  <div>
    <AppView
      repos={Object {}}
      viewKey="someViewKey"
    >
      <div />
    </AppView>
  </div>
</AppContainer>
`;

exports[`AppContainer Initialization should trigger componentwillmount life cycle event 1`] = `
<AppContainer
  refreshRepo={
    [MockFunction] {
      "calls": Array [
        Array [
          "whatever",
          "someKey",
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  repos={Object {}}
  viewKey="someViewKey"
>
  <div>
    <AppView
      repos={Object {}}
      viewKey="someViewKey"
    >
      <div />
    </AppView>
  </div>
</AppContainer>
`;

exports[`AppContainer When view key is VK_ADD_NEW_REPO should render an add new repo view 1`] = `
<div>
  <AppView
    repos={Object {}}
    viewKey="add_new_repo"
  />
</div>
`;

exports[`AppContainer When view key is VK_REPOS should render a status view 1`] = `
<div>
  <AppView
    repos={Object {}}
    viewKey="repos"
  />
</div>
`;
